### DISCUSSION TITLE
Message acknowledgement with external subsystems

### DISCUSSION THREAD

**Discussion by beatrausch** [nats-io/nats.go#907] [Q&A]
Hi,

I am looking for a robust way to acknowledge a message after being asynchronously processed by a subsystem e.g. some external library that is not under my control, thus I cannot pass the Msg and call Ack.

The current approach passes the reply string to the subsystem. When the subsystem responses I simply call Publish on the nats connection with either +ACK or -ACK. Unfortunately, ackAck and ackNak are private in the nats go client, thus it feels a bit hacky.

I am interested in your opinion about that solution and if there are plans to support such a use case in a simpler way e.g. by adding additional methods Ack, Nak, Term, etc. in the Jetstream interface.

Regards

**Comment by derekcollison** [comment]
They are all present on a received message, can you hold onto the message and when the subsystem responds just call msg.Ack() or msg.Nak()?

**Reply by beatrausch** [reply]
Unfortunately, I cannot hold the current message and wait until the subsystem responses. Due to the asynchronous, distributed design, it is not guaranteed that the acknowledgement is sent by the same goroutine which receives the Msg from NATS and to make more complicated it might not be the same process.

**Reply by derekcollison** [reply]
You could construct a nats.Msg on the fly with the reply subject and a subscription bound to the connection you want to use. Then you could use m.Ack() etc.

**Reply by kozlovic** [reply]
@derekcollison But that won't work since we check that the message is associated with a subscription, that the subscription is "JetStream" one. That's also how we get the connection. With a "bare" new message, there is no way Ack() can work.

**Reply by derekcollison** [reply]
Hmm, ok. Fair point. So I guess they will need to recreate all of the Ack calls for situations that just have a NATS connection and reply subject.

**Reply by beatrausch** [reply]
By creating a new nats.Msg from an Subscription, that comes from a JetStream, Ack calls on the newly created message works.

**Comment by tpihl** [comment]
How do you handle retries or multiple processing of the same info.  I understand that the async subsystem is a black box but I don’t understand the solution design. Keeping the msg in a blocked goroutine until processing is done would make it possible to use acks the simple way (including extending and nak). 

If you pick up the completion via some other means (webhook or similar), a normal nats request/reply would be a great way to “find” the right goroutine.  Or if there is many many requests, get the nats msg from a map or something and then ack it. 

Either way, mechanisms for idempotemcy and retry/fallback actions should also be considered (and that can be very hard to do unless “someone” holds the active msg.

**Reply by beatrausch** [reply]
_How do you handle retries or multiple processing of the same info_

Since at-least-once semantic in the data processing is fine, there is no special processing required. Furthermore, each message has in its payload a unique id that allows the downstream processing to identify messages that haven delivered multiple times into the subsystem. 


_I understand that the async subsystem is a black box but I don’t understand the solution design. Keeping the msg in a blocked goroutine until processing is done would make it possible to use acks the simple way (including extending and nak)._

That’s absolutely true, but as I took over, the application design, was mostly done. May be that’s something that I could consider during future changes.
